contracts/migration/LiquidityMigrator.sol

```diff
@@ -2,92 +2,110 @@
+        uint256 lpAmount = amount < max ? amount : max;
+        oldTreasury.manage(address(pair), lpAmount);
 
-        // Remove the V1 liquidity
-        IERC20(pair).approve(address(routerFrom), oldLPAmount);
-        (uint256 amountToken, uint256 amountOHM) = routerFrom.removeLiquidity(
-            token,
+        // Remove the liquidity
+        pair.approve(address(router), lpAmount);
+        router.removeLiquidity(
+            address(token),
             address(oldOHM),
-            oldLPAmount,
-            _minA,
-            _minB,
+            lpAmount,
+            minToken,
+            minOhm,
             address(this),
-            _deadline
+            deadline
         );
+        
+        // deposit the paired token into the new treasury
+        uint256 amountToken = token.balanceOf(address(this));
+        token.approve(address(newTreasury), amountToken);
+        newTreasury.deposit(amountToken, address(token), newTreasury.tokenValue(address(token), amountToken));
+
+        // burn the OHM
+        oldOHM.burn(oldOHM.balanceOf(address(this)));
+    }
 
-        // Migrate the V1 OHM to V2 OHM
-        oldOHM.approve(address(migrator), amountOHM);
-        migrator.migrate(amountOHM, IMigrator.TYPE.UNSTAKED, IMigrator.TYPE.UNSTAKED);
-        uint256 amountNewOHM = newOHM.balanceOf(address(this)).sub(initialNewOHMBalance); // # V1 out != # V2 in
+    function migrate(
+        uint256 amount,
+        IERC20 pair,
+        IUniswapV2Router router,
+        IERC20 token,
+        uint256 minToken,
+        uint256 minOhm,
+        uint256 deadline
+    ) external onlyGovernor {
+        // Fetch the treasury balance of the given liquidity pair
+        uint256 max = pair.balanceOf(address(oldTreasury));
+        uint256 lpAmount = amount < max ? amount : max;
+        oldTreasury.manage(address(pair), lpAmount);
 
-        // Add the V2 liquidity
-        IERC20(token).approve(address(routerTo), amountToken);
-        newOHM.approve(address(routerTo), amountNewOHM);
-        routerTo.addLiquidity(
-            token,
-            address(newOHM),
-            amountToken,
-            amountNewOHM,
-            amountToken,
-            amountNewOHM,
-            address(newTreasury),
-            _deadline
+        // Remove the liquidity
+        pair.approve(address(router), lpAmount);
+        router.removeLiquidity(
+            address(token),
+            address(oldOHM),
+            lpAmount,
+            minToken,
+            minOhm,
+            address(this),
+            deadline
         );
+        
+        // deposit the paired token into the new treasury
+        uint256 amountToken = token.balanceOf(address(this));
+
+        // burn the old OHM
+        uint256 amountOld = oldOHM.balanceOf(address(this));
+        oldOHM.burn(amountOld);
+
+        // mint new OHM after converting to new supply
+        uint256 amountNew = amountOld * newsohm.index() / oldsohm.index();
```

Comment:
Instead of having this contract mint new ohm, we accomplished this by calling the migrator. Keeps the logic / bookkeeping all in one place. What's the reasoning for doing the accounting in this contract?

Reply:
Also needs safemath

Reply:
the migrator does not have corresponding gOHM for the tokens in this pool afaik. verified at the time of writing but can be checked with: gOHM in migrator * index vs (sOHMv1 circulating * 46.7) + wsOHM circulating + OHMv1 circulating. intent is to avoid a deficit in that pool, as staked supply was used for the number of gOHM to mint at migration.